src/internal/Internal/Invariant.hs

```diff
@@ -0,0 +1,33 @@
+-- |
+-- Copyright: © 2018-2020 IOHK
+-- License: Apache-2.0
+--
+-- Provides internal functions relating to verification of invariants.
+--
+module Internal.Invariant
+    ( invariant
+    ) where
+
+import Prelude
+
+-- | Checks whether or not an invariant holds, by applying the given predicate
+--   to the given value.
+--
+-- If the invariant does not hold (indicated by the predicate function
+-- returning 'False'), throws an error with the specified message.
+--
+-- >>> invariant "not empty" [1,2,3] (not . null)
+-- [1, 2, 3]
+--
+-- >>> invariant "not empty" [] (not . null)
+-- *** Exception: not empty
+invariant
```

Comment:
this is third time I was considering if this should not be some time in the future in `cardano-aux` or `cardano-prelude` if we use it

Reply:
> this is third time I am thinking loudly if this should not be some time in the future in cardano-aux or cardano-prelude if we use it

I think you're right. Though I'll leave that one for a future PR. :smile_cat: